src/mozanalysis/experiment.py

```diff
@@ -0,0 +1,509 @@
+
+        return enrollments
+
+    def get_per_client_data(
+        self, enrollments, df, metric_list, last_date_full_data, conv_window_start_days,
+        conv_window_length_days, keep_client_id=False
+    ):
+        """Return a DataFrame containing per-client metric values.
+
+        Args:
+            enrollments: A spark DataFrame of enrollments, like the one
+                returned by `self.get_enrollments()`.
+            df: A spark DataFrame containing the data needed to calculate
+                the metrics. Could be `main_summary` or `clients_daily`.
+                _Don't_ use `experiments`; as of 2019/04/02 it drops data
+                collected after people self-unenroll, so unenrolling users
+                will appear to churn. Must have at least the following
+                columns:
+                    - client_id (str)
+                    - submission_date_s3 (str)
+                    - data columns referred to in `metric_list`
+                Ideally also has:
+                    - experiments (map)
+            metric_list: A list of columns that aggregate and compute
+                metrics, e.g.
+                `[F.coalesce(F.sum(df.metric_name), F.lit(0)).alias('metric_name')]`
+            last_date_full_data (str): The most recent date for which we
+                have complete data, e.g. '20190322'. If you want to ignore
+                all data collected after a certain date (e.g. when the
+                experiment recipe was deactivated), then do that here.
+            conv_window_start_days (int): the start of the conversion window,
+                measured in 'days since the user enrolled'. We ignore data
+                collected outside this conversion window.
+            conv_window_length_days (int): the length of the conversion window,
+                measured in days.
+            keep_client_id (bool): Whether to return a `client_id` column.
+                Defaults to False to reduce memory usage of the results.
+
+        Returns:
+            A spark DataFrame of experiment data. One row per `client_id`.
+            One or two metadata columns, then one column per metric in
+            `metric_list`. Then one column per sanity-check metric.
+            Columns:
+                - client_id (str, optional): Not necessary for
+                    "happy path" analyses.
+                - branch (str): The client's branch
+                - [metric 1]: The client's value for the first metric in
+                    `metric_list`.
+                - ...
+                - [metric n]: The client's value for the nth (final)
+                    metric in `metric_list`.
+                - [sanity check 1]: The client's value for the first
+                    sanity check metric.
+                - ...
+                - [sanity check n]: The client's value for the last
+                    sanity check metric.
+
+            This format - the schema plus there being one row per
```

Comment:
Do we have some artifact of this decision?

Reply:
Heh, I was being a little cheeky here. Now (i.e. before this PR is merged) is an appropriate time to re-open the discussion if you would like to. [Here's the doc I took to the team meeting](https://docs.google.com/document/d/1S5_j6T-buAlQXqsevgGUvAD2Xt8qjeEjHfwc9xGhjiw/edit#heading=h.l1m5icicw80k) - we had a discussion and people seemed OK with it though there was no vote. If you or anyone else has concerns (including with the process) then please do speak up before next Wed; we should settle this issue before merging this code and text.

Reply:
Sounds good. We could also use this PR thread as the artifact of that decision if we are sure to get sguha and rweiss's explicit sign off as well.

Reply:
You can consider this my explicit buy-in, fwiw.

Reply:
this looks good to me. I'm curious however, where does data cleaning(trimming/capping/cleaning) happen? is this something the analyst does before this spark dataframe of experiment data is returned(i.e so that the input to this function will be already cleaned)

Reply:
If you're tidying up individual rows of the data source (e.g. individual pings) then you'd need to do that before passing it to `get_per_client_data`. If you're capping or filtering outlier clients based on their values for a metric, then you'd do that with the output of `get_per_client_data`.

Does that sound reasonable and cover your usecases, or is there one I missed?

Reply:
This addresses improving analysis efficiency for data science, so r+ from me.  I'll defer to @ethompso28 on whether any further review is needed (e.g. from data eng).